### ISSUE TITLE
[FEATURE] Data Deduplication Support

### ISSUE THREAD

**Issue by netwarex** [longhorn/longhorn#5101]
## Is your feature request related to a problem? Please describe (👍 if you like this request)

There are volumes which containing repetitive data, and I would like to have the option for those volumes to have deduplication. This would not only save space, but improves performance when replicating those volumes.

## Describe the solution you'd like

To allow deduplication, the most practical way is on a filesystem level, so we can choose eg.: btrfs or zfs

## Describe alternatives you've considered

Second option is a block level deduplication, which can be done by longhorn itself, however this would need much more effort.

## Additional context

This would be really useful for CI/CD use cases where the output is and logs are really similar, but can grow fast.

**Comment by derekbit** [comment]
Yeah, as you mentioned, you can use zfs supporting deduplication for now, but we didn't check its stabability in Longhorn for now. What you request is related to [sharding support](https://github.com/longhorn/longhorn/issues/1061). After implementing the feature, the deduplication is theoretically doable.

**Comment by netwarex** [comment]
Of course I would use something that is supported ;) Thanks for the feedback!

**Comment by derekbit** [comment]
Besides of ext4 and xfs, we can check the zfs which is also a popular filesystem.
cc @innobead @shuo-wu @PhanLe1010

**Comment by netwarex** [comment]
@derekbit What about XFS's reflink support. Is this support with Longhorn?

**Comment by innobead** [comment]
> @derekbit What about XFS's reflink support. Is this supported by Longhorn?

Longhorn 1.4.0 supports `mkfsParams` setting for storage class, so you can customize the mkfs.xfs options. Go check https://longhorn.io/docs/1.4.0/volumes-and-nodes/create-volumes/#creating-longhorn-volumes-with-kubectl

**Comment by pchang388** [comment]
This is just one datapoint but I am running proxmox server with a LSI 9300-16i (IT mode) with 4x 1.92TB Intel D3-4510 with zfs (128k record size and 16k vol block size) on hardware:

- ASUS w680 Pro ACE motherboard
- Intel i9 13900k
- DDR5 ECC RAM 4800

My current workloads are very simple as shown in the screenshot but everything seems to be working fine with default settings (only increased some timeout values to reduce flapping/rebuilding during node restarts). I will say though zfs performance is tricky for consumer/homelab set ups and without a proper set up (dram and plp on SSDs, HBA/IT controller, etc.), your performance will suffer and may lead users to think it's an issue with longhorn + zfs but may just be an issue with zfs + hardware. If you have the right tuning (ashift, atime, record size, block size, etc.) and hardware, zfs can be very performant and reliable.

<img width="1761" alt="Screenshot 2023-04-19 at 11 00 39 AM" src="https://user-images.githubusercontent.com/51681873/233117538-2fe008c8-a167-416f-961c-b8dd0cb40dbe.png">

**Comment by Berodin** [comment]
Is there any focus on this topic? Due to.... reasons data duplication and more efficient use of Disks gained some velocity in the last months.

**Comment by derekbit** [comment]
> Is there any focus on this topic? Due to.... reasons data duplication and more efficient use of Disks gained some velocity in the last months.

It is not in our plan for now. 

>  data duplication and more efficient use of Disks gained some velocity in the last months.

Do you have many snapshots? We have some methods to reduce the space consumption
https://longhorn.io/kb/manual-recovery-of-nodes-with-insufficient-space/
